cpp/src/round/round.cu

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/round.hpp>
+#include <cudf/round.hpp>
+#include <cudf/scalar/scalar.hpp>
+// #include <cudf/scalar/scalar_factories.hpp>
+// #include <cudf/utilities/error.hpp>
+// #include <cudf/utilities/traits.hpp>
+// #include <cudf/fixed_point/fixed_point.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace cudf {
+
+namespace detail {
+
+struct round_fn {
+  template <typename T, typename... Args>
+  std::enable_if_t<not cudf::is_floating_point<T>(), std::unique_ptr<column>> operator()(
+    Args&&... args)
+  {
+    CUDF_FAIL("fail for the moment");
+  }
+
+  template <typename T, typename... Args>
+  std::enable_if_t<cudf::is_floating_point<T>(), std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t scale,
+    cudf::round_option round,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto result   = cudf::make_fixed_width_column(input.type(), input.size());
+    auto out_view = result->mutable_view();
+    auto const n  = static_cast<int32_t>(std::pow(10, std::abs(scale)));
+
+    if (scale == 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [] __device__(T e) -> T { return std::roundf(e); });
```

Comment:
fyi, I wouldn't recommend trying to use a `nvstd::function` here. Foreign function calls in device code can have pretty high overhead.

Reply:
I will give it a shot. Not sure if you saw the last post on the `cdd-cuda` thread but the non-empty capture lists make them hard (impossible?) to combine

Reply:
I don't think this does what we want for all floating point values.  FLOAT64 would have to be cast to a float for `roundf`.  I think in that case we would want just `round`